### PR TITLE
feat(instrumentation-google-generativeai): add OTel 1.40 GenAI semantic conventions instrumentation

### PR DESCRIPTION
**Pull request by lenatraceloop**
Summary
                                                                                                                                                                                                                                              
  - Creates new @traceloop/instrumentation-google-generativeai — OpenTelemetry instrumentation for the @google/genai SDK, implementing OTel GenAI semantic conventions v1.40
  - Adds mapGenAIContentBlock to @traceloop/instrumentation-utils — a shared content block mapper for Google GenAI Part types (text, reasoning, blob, uri, tool_call, tool_call_response, executable_code, code_execution_result)             
  - Registers the instrumentation in @traceloop/node-server-sdk                                                                                                                                                                               
                                                                                                                                                                                                                                              
  Attributes captured (aligned with OTel 1.40 upstream):                                                                                                                                                                                      
  - gen_ai.request.stop_sequences — stored as native string[]
  - gen_ai.usage.cache_read.input_tokens — from usageMetadata.cachedContentTokenCount                                                                                                                                                         
  - gen_ai.usage.reasoning_tokens — from usageMetadata.thoughtsTokenCount (thinking models)
  - gen_ai.request.thinking.budget_tokens — from thinkingConfig.thinkingBudget                                                                                                                                                                
                                                                                                                                                                                                                                              
  Test coverage: 100% — 131 tests across 4 test files:
  - instrumentation.test.ts — end-to-end span attribute coverage (generateContent + streaming, traceContent on/off, tool use, system instructions, thinking config, token usage)                                                              
  - content_parts.test.ts — all Part types including edge cases (video modality, undefined response, string response)                                                                                                                         
  - finish_reasons.test.ts — all Gemini finish reason mappings for both span-level and per-message values
  - semconv.test.ts — all OTel 1.40 attribute constants                                                                                                                                                                                       
                  
  Fixes TLP-2077                                                                                                                                                                                                                              
                  
  Test plan                                                                                                                                                                                                                                   
   
  - pnpm nx test @traceloop/instrumentation-google-generativeai — 131 tests passing                                                                                                                                                           
  - pnpm nx build @traceloop/instrumentation-google-generativeai — builds cleanly
  - pnpm nx run-many -t build — all 17 packages build
  - pnpm nx run-many -t test — all tests pass (qdrant requires a running Qdrant instance, pre-existing)         

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Added OpenTelemetry instrumentation for Google Generative AI with automatic spans, attribute mapping, content tracing toggle, and optional exception logging.
  * Added content-block mapping and standardized finish-reason mapping (includes unspecified → empty string).
  * Integrated Google GenAI instrumentation into Traceloop SDK with manual/automatic activation.

* **Documentation**
  * Added README with setup and usage examples for the new instrumentation.

* **Tests**
  * Added extensive tests covering formatting, finish reasons, streaming, span behavior, and error handling.

* **Chores**
  * Added build, bundling, lint, and TypeScript configs plus packaging updates and tooling ignores.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

### CODE REVIEW

File: packages/instrumentation-google-generativeai/README.md [hunk 1]
```diff
@@ -0,0 +1,55 @@
+# OpenTelemetry Google Gen AI SDK instrumentation for Node.js
+
+[![NPM Published Version][npm-img]][npm-url]
+[![Apache License][license-image]][license-image]
```
[the remark this answers is not in the document]
Comment (lenatraceloop) [reply packages/instrumentation-google-generativeai/README.md]: Both fixes are now applied in the latest commit:                                                                                                                                                                                            
  1. License badge now links to [license-url] instead of [license-image]
  2. NPM badge slug corrected to instrumentation-google-generativeai                                                                                                                                                                          
                                                                                                                                                                                                                                              
  Regarding the "not found" on the npm badge — the package hasn't been published yet, it will resolve automatically after the first release.

File: packages/instrumentation-google-generativeai/src/instrumentation.ts [hunk 2]
```diff
@@ -0,0 +1,628 @@
...
+  IMAGE_OTHER: FinishReasons.ERROR,
+};
+
+/**
+ * Maps a Gemini finish reason string to the value used in gen_ai.response.finish_reasons.
+ * Follows OTel passthrough convention: lowercase, strip the "finish_reason_" prefix.
+ * IMAGE_* and other vendor-specific prefixes are intentionally retained (lowercased only),
+ * as they are not OTel well-known values but are valid vendor-specific extensions.
+ * Used only for the span-level array attribute (not per-message finish_reason).
+ */
```
Comment (nina-kollman) [inline packages/instrumentation-google-generativeai/src/instrumentation.ts]: It cannt be that the doc is longer than the function itself

File: packages/instrumentation-google-generativeai/src/instrumentation.ts [hunk 3]
```diff
@@ -0,0 +1,628 @@
...
+  private _shouldSendPrompts(): boolean {
+    const ctxValue = context.active().getValue(CONTEXT_KEY_ALLOW_TRACE_CONTENT);
+    if (ctxValue !== undefined) return ctxValue as boolean;
+    return this._config.traceContent !== undefined
+      ? this._config.traceContent
+      : true;
+  }
+
+  private wrapGoogleGenAI() {
+    // eslint-disable-next-line @typescript-eslint/no-this-alias
```
Comment (nina-kollman) [inline packages/instrumentation-google-generativeai/src/instrumentation.ts:139]: Why is the eslint disable needed in this file?

File: packages/instrumentation-google-generativeai/src/types.ts [hunk 4]
```diff
@@ -0,0 +1,6 @@
+import { InstrumentationConfig } from "@opentelemetry/instrumentation";
+
+export interface GenAIInstrumentationConfig extends InstrumentationConfig {
```
Comment (nina-kollman) [inline packages/instrumentation-google-generativeai/src/types.ts:3]: Is this in use?

File: packages/instrumentation-utils/src/content-block-mappers.ts [hunk 5]
```diff
@@ -257,6 +257,135 @@ export function mapOpenAIContentBlock(block: any): object {
   }
 }
 
+// =============================================================================
+// Google Gen AI (@google/genai)
```
Comment (nina-kollman) [inline packages/instrumentation-utils/src/content-block-mappers.ts]: I have a general question about this file. I can see mappers for different instrumentations. Why are the functions here in the utils and not in the specific instrumentation folder. 
This folder is for function that can be shared across instrumentations

File: packages/instrumentation-google-generativeai/src/instrumentation.ts (hunk 2 above)
Comment (lenatraceloop) [reply packages/instrumentation-google-generativeai/src/instrumentation.ts]: Fixed

File: packages/instrumentation-google-generativeai/src/instrumentation.ts (hunk 3 above)
Comment (lenatraceloop) [reply packages/instrumentation-google-generativeai/src/instrumentation.ts:139]: This pattern is used consistently across all instrumentations in this codebase — it's the standard way wrappers are written here.

File: packages/instrumentation-google-generativeai/src/types.ts (hunk 4 above)
Comment (lenatraceloop) [reply packages/instrumentation-google-generativeai/src/types.ts:3]: Yes, it's used — _config.traceContent and _config.exceptionLogger. The type declaration is needed so TypeScript knows about these custom fields.

File: packages/instrumentation-utils/src/content-block-mappers.ts (hunk 5 above)
Comment (lenatraceloop) [reply packages/instrumentation-utils/src/content-block-mappers.ts]: Fixed. moved to local content-block-mappers.js

Comment (nina-kollman) [reply packages/instrumentation-utils/src/content-block-mappers.ts]: I can still see it here

File: packages/traceloop-sdk/package.json [hunk 6]
```diff
@@ -100,6 +101,7 @@
     "@anthropic-ai/sdk": "^0.80.0",
     "@aws-sdk/client-bedrock-runtime": "^3.969.0",
     "@google-cloud/aiplatform": "^4.4.0",
+    "@google/genai": "^1.49.0",
```
Comment (OzBenSimhonTraceloop) [inline packages/traceloop-sdk/package.json]: ### 
```suggestion
    "@google/genai": "^1.50.1",
```
lets use latest

**Review by OzBenSimhonTraceloop — COMMENTED** [review]
P1 — Must Fix
1. Span-level finish_reasons uses passthrough lowercase instead of OTel mapped values
[instrumentation.ts:88-91](vscode-webview://0m585ttf9nt5bbq2nfcbakeim4ba3daacorh1klr9ign2g9d1feg/packages/instrumentation-google-generativeai/src/instrumentation.ts#L88-L91), [instrumentation.ts:468-476](vscode-webview://0m585ttf9nt5bbq2nfcbakeim4ba3daacorh1klr9ign2g9d1feg/packages/instrumentation-google-generativeai/src/instrumentation.ts#L468-L476)

geminiReasonToSpanValue() lowercases vendor values (SAFETY → "safety", MAX_TOKENS → "max_tokens"). Should use genaiFinishReasonMap like Anthropic/OpenAI instrumentations do, producing OTel canonical values ("content_filter", "length"). Traceloop Python uses one mapper for both paths — no dual-path design. Delete geminiReasonToSpanValue, use genaiFinishReasonMap with ?? r fallback.

2. Streaming accumulation merges thinking + regular text
[instrumentation.ts:293-302](vscode-webview://0m585ttf9nt5bbq2nfcbakeim4ba3daacorh1klr9ign2g9d1feg/packages/instrumentation-google-generativeai/src/instrumentation.ts#L293-L302)

Consecutive text parts concatenated without checking thought flag. A {text: "...", thought: true} followed by {text: "..."} merges into one reasoning part. Fix: add && !!prev.thought === !!part.thought to the concat condition.

P2 — Should Fix
3. OTHER maps to "other" — should be "error" (cross-SDK drift)
[instrumentation.ts:74](vscode-webview://0m585ttf9nt5bbq2nfcbakeim4ba3daacorh1klr9ign2g9d1feg/packages/instrumentation-google-generativeai/src/instrumentation.ts#L74)

Traceloop Python and upstream OTel Python both map OTHER → "error". JS maps to "other" which isn't in the OTel FinishReason enum. Fix: OTHER: FinishReasons.ERROR.

4. BlobPart/UriPart modality omitted when mimeType absent
[content-block-mapper.ts:72-78](vscode-webview://0m585ttf9nt5bbq2nfcbakeim4ba3daacorh1klr9ign2g9d1feg/packages/instrumentation-google-generativeai/src/content-block-mapper.ts#L72-L78)

modality is required per OTel schema. Fall back to "document" when mimeType is unknown.

Cross-SDK Drift (Python vs JS)
5. Provider name: gcp.gen_ai (Python) vs gcp.gemini (JS)
Both valid OTel values but must agree. gcp.gemini is more specific for AI Studio API. Align across SDKs.

6. Top-level finish_reasons — per-candidate alignment vs dedup
Python preserves per-candidate alignment (no dedup, "" placeholders, omits attr only when all empty). JS deduplicates via new Set() and drops nulls. Align on Python's approach — array length should match candidate count.

P3 — Nice to Have
7. Missing test: per-message finish_reason in multi-candidate output — test at line 597 checks span attr but not outputMessages[*].finish_reason.

8. Missing test: streaming + thinking parts — would catch P1 #2.

9. Missing test: streaming where final chunk has null finishReason — verify span attr omitted and output message has finish_reason: "".

10. Empty string filter on span finish_reasons — filter((r) => r != null) doesn't exclude "". Add && r !== "" defensively.

File: packages/traceloop-sdk/package.json (hunk 6 above)
Comment (lenatraceloop) [reply packages/traceloop-sdk/package.json]: fixed

**Comment by lenatraceloop** [comment]
Thanks for the detailed review! All issues have been addressed:

  P1 — Fixed:
  1. Deleted geminiReasonToSpanValue, now using genaiFinishReasonMap[r] ?? r everywhere — produces OTel canonical values ("content_filter", "length", etc.)
  2. Added && !!prev.thought === !!part.thought to the streaming concat condition — thinking and regular text parts are now kept separate

  P2 — Fixed:
  3. OTHER now maps to FinishReasons.ERROR — aligned with Python and upstream OTel
  4. modality is now always set on BlobPart/UriPart, falling back to "document" when mimeType is absent
  5. Provider name changed from "gcp.gemini" to "gcp.gen_ai" — aligned with Python instrumentation and OTel contrib
  6. Removed new Set() dedup, finish_reasons array is now per-candidate with "" placeholders, attribute omitted only when all values are empty — matches Python behavior exactly

  P3 — Done:
  7. Added finish_reason assertions to the multi-candidate output messages test
  8. Added streaming + thinking parts test
  9. Added streaming with null finishReason test — verifies span attr is omitted and finish_reason: ""
  10. Resolved via the .some(r => r !== "") condition

File: packages/instrumentation-google-generativeai/src/instrumentation.ts [hunk 7]
```diff
@@ -0,0 +1,615 @@
...
+  SPII: FinishReasons.CONTENT_FILTER,
+  MALFORMED_FUNCTION_CALL: FinishReasons.ERROR,
+  IMAGE_SAFETY: FinishReasons.CONTENT_FILTER,
+  FINISH_REASON_UNSPECIFIED: FinishReasons.FINISH_REASON_UNSPECIFIED,
+  UNEXPECTED_TOOL_CALL: FinishReasons.ERROR,
+  IMAGE_PROHIBITED_CONTENT: FinishReasons.CONTENT_FILTER,
+  NO_IMAGE: FinishReasons.ERROR,
+  IMAGE_RECITATION: FinishReasons.CONTENT_FILTER,
+  IMAGE_OTHER: FinishReasons.ERROR,
+};
```
[the remark this answers is not in the document]
Comment (lenatraceloop) [reply packages/instrumentation-google-generativeai/src/instrumentation.ts:86]: Thanks for the suggestion. I checked this against the Python implementation, which is the source of truth for this normalization logic.

In Python, the mapping is intentionally:

* `FINISH_REASON_UNSPECIFIED` -> `""`
* `OTHER` -> `"error"`
* `MALFORMED_FUNCTION_CALL` / `UNEXPECTED_TOOL_CALL` / `NO_IMAGE` / `IMAGE_OTHER` -> `"error"`

Because of that, changing `OTHER` to `FINISH_REASON_UNSPECIFIED` in JS would make the behavior diverge from Python.

So this change is not needed.

File: packages/ai-semantic-conventions/src/SemanticAttributes.ts [hunk 8]
```diff
@@ -162,4 +162,6 @@ export const FinishReasons = {
   TOOL_CALL: "tool_call",
   CONTENT_FILTER: "content_filter",
   ERROR: "error",
+  OTHER: "other",
```
Comment (OzBenSimhonTraceloop) [inline packages/ai-semantic-conventions/src/SemanticAttributes.ts]: really needed?

**Review by OzBenSimhonTraceloop — APPROVED** [review]

File: packages/ai-semantic-conventions/src/SemanticAttributes.ts (hunk 8 above)
Comment (lenatraceloop) [reply packages/ai-semantic-conventions/src/SemanticAttributes.ts]: removing it
